[Backend] Fix incorrect shared layout for dot operands rank==3

AlexAUT opened the pull request:
#4904 moved the layout rewrite for dot operands from `blocked->mma` to` blocked->shared->mma` for `wmma` from [`ReduceDataDeplucation`](https://github.com/triton-lang/triton/blob/main/lib/Dialect/TritonGPU/Transforms/ReduceDataDuplication.cpp#L54-L79) to `DecomposeUnsupportedConversion` with [this change](https://github.com/triton-lang/triton/pull/4904/files#diff-764fa3852edb91756d5593808fb2cea98792e8b8bf7f76e492d5af8fc599fcb3R47-R53). However, `DecomposeUnsupportedConversion` was missing the special case for `rank==3` which is copied over in this PR.

lezcano reviewed (commented):
The changes to `wmma` were not intentional, it was land race with https://github.com/triton-lang/triton/pull/4538. Feel free to revert the change making `ReduceDataDuplicaton`'s condition apply only to Ampere.

That being said, would it make sense here  to simply use `dstOrder` as the order to be used, similar to how it's done in
https://github.com/triton-lang/triton/blob/692143cd869f2fa1501edb1db76fb452b85ac914/lib/Analysis/Allocation.cpp#L124
This one already gives you the order you want for `rank=3`, while the current one could end up in a funny state where the order does not match that of the input or the output.

AlexAUT commented:
Thanks for the feedback. I reverted the change in `ReduceDataDuplication`.

I also switched to `getThreadOrder(dstEncoding)`. We want to use that for all ranks right? Because for `rank < 3` source and destination order are the same anyways?

lezcano commented on lib/Conversion/TritonGPUToLLVM/DecomposeUnsupportedConversions.cpp:
```diff
@@ -99,15 +99,21 @@ void decomposeBlockedToDotLayoutConversion(ModuleOp module) {
+      auto srcOrder = srcBlocked.getOrder();
+      auto rank = srcOrder.size();
+      SmallVector<unsigned> sharedOrder;
+      if (rank == 3) {
+        sharedOrder = gpu::getThreadOrder(dstEncoding);
+      } else {
+        sharedOrder = srcOrder;
+      }
```
If you are going this route, you probably want to do it for all ranks, otherwise this heuristic would be incredibly counterintuitive.

AlexAUT replied:
But then we might change the shared layout order for `rank != 3` with this change? I can also revert all changes except for the condition in `ReduceDataDuplication` to make it ampere specific. Then we have the same behavior as before #4904.

lezcano replied:
I just find very weird the current behaviour for `rank==3`. @Jokeren thoughts?

Jokeren replied:
Actually block to dot decomposition can be deprecated soon
https://github.com/triton-lang/triton/compare/main...keren/dot-mma#diff-30fb59df648f6c4ec5db24c51ef728a8d56104e151bd52e8d482b6951c207aa7R95

Jokeren replied:
I agree the special condition is weird. Can you attach a test case for us to take a look at?

AlexAUT replied:
When you run `python/test/unit/language/test_core.py::test_dot3d[1-1-64-64-64-32-32-float16-float32]` it will trigger this [assert](https://github.com/triton-lang/triton/blob/main/lib/Conversion/TritonGPUToLLVM/MemoryOpToLLVM.cpp#L27-L29). This happens when `wmma` is used for the dot.

AlexAUT commented:
#4950 changed the condition to restore the old behavior for `wmma` so this problem is fixed now. Thanks @lezcano.
